[nativert] move intrusive list to c10/util

dolpm opened the pull request:
Summary:
nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff moves intrusive list to c10/util

Test Plan: CI

Differential Revision: D74104595

Skylion007 commented on c10/util/IntrusiveList.h:
```diff
@@ -0,0 +1,189 @@
+  template <typename T>
+  friend class IntrusiveList;
+
+  IntrusiveListHook* next_;
```
nit: initialize as nullptr?

Skylion007 commented on c10/util/IntrusiveList.h:
```diff
@@ -0,0 +1,189 @@
+    return iterator{&n};
+  }
+
+  size_t size() const {
```
If we aren't implementing size, the `empty()` should at least be defined so can easily check if the list is empty

cyyever reviewed (approved):
If linters happy, I'm happy.

dolpm committed:
[nativert] move intrusive list to c10/util (#152754)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152754

nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff moves intrusive list to c10/util

Test Plan: CI

Differential Revision: D74104595

Skylion007 commented:
As @albanD point out, why don't we just point this to a regular stdlib::list?

zhxchen17 commented:
> What prevents us from using std list for this?

@albanD @Skylion007 IntrusiveList is different from std::list because it allows us to allocate all the memory at once for a node, while std::list forces us to heap allocate twice for list entry and the node itself separately. In addition to memory efficiency and performance concerns, this also allows to track nodes safely without passing around the list object (because intrusive list maintains a stronger invariant that expiration will always cause unlinking).

albanD commented:
Would https://github.com/pytorch/pytorch/blob/main/c10/util/SmallVector.h help then to inline as much as possible?

swolchok commented:
> forces us to heap allocate twice for list entry and the node itself separately.

A quick check of both the libc++ and libstdc++ list implementations seems to suggest that this is not the case. Why do you think two heap allocations are required?
